### DISCUSSION TITLE
file_programmer.py

### DISCUSSION THREAD

**Discussion by xiongwei0714** [pyocd/pyOCD#1436] [Ideas]
Can the developer add the program details as below to the return value of file_programmer.FileProgramer.program?
"Erased %d bytes (%s), programmed %d bytes (%s), skipped %d bytes (%s) at %.02f kB/s" ;
or "Erased chip, programmed %d bytes (%s), skipped %d bytes (%s) at %.02f kB/s" ;
or values used in above string;
or values can be used to calculate the value in above string.

I know the details are logged by the loader.py, but  it's not easy to get the details from log. I need the details to be displayed on my GUI.
I know I can implment the sub class and rewrite some function to achieve my goal, but I more hope I can get the details from file_programmer.FileProgramer.program.

**Comment by xiongwei0714** [comment]
It would be better to add the progress callback for eraser.FlashEraser.erase and return the details for erased size and speed like the programmer above.

**Comment by flit** [comment]
(Sorry for the very late reply; I was sick for quite a long time.)

The best way to implement this would probably be to return a data class containing the flash+erase statistics from `FileProgrammer.program()`.

And yes, I already have a todo note to add the progress callback for `FlashEraser`.

Fyi, the best way to get these features implemented is to submit a pull request yourself (eg, pyocd is typical open source project). Otherwise there is a long list of higher priority items that need attention.
